src/AutoSchedule.cpp

```diff
@@ -0,0 +1,2312 @@
+
+    estimates[inner_name] = factor;
+    estimates[outer_name] =
+            std::ceil((float)estimates.at(arg_name)/factor);
+    estimates.erase(arg_name);
+
+    return make_pair(inner, outer);
+}
+
+void vectorize_stage(Stage f_handle, Definition def, Function func,
+                     const Target &t, set<string> &rvars,
+                     map<string, int> &estimates, string &sched) {
+    const vector<Dim> &dims = f_handle.get_schedule().dims();
+    int vec_dim_index = -1;
+
+    // Set the vector length as the maximum of the values produced by a
+    // function
+    int vec_len = 0;
+    for (auto &type: func.output_types()) {
+        vec_len = std::max(vec_len, t.natural_vector_size(type));
+    }
+
+    for (int d = 0; d < (int) dims.size() - 1; d++) {
+        string dim_name = get_base_name(dims[d].var);
+        bool can_vectorize = true;
+        if (rvars.find(dim_name) != rvars.end()) {
+            can_vectorize = can_parallelize_rvar(dim_name, func.name(), def);
+        }
+        if (estimates.find(dim_name) != estimates.end()) {
+            if (can_vectorize && estimates[dim_name] >= vec_len) {
+                vec_dim_index = d;
+                break;
+            }
+        }
+    }
+
+    if (vec_dim_index >= 0) {
+        string vec_dim_name = get_base_name(dims[vec_dim_index].var);
+        Var vec_var(vec_dim_name);
+
+
+        bool is_rvar = (rvars.find(vec_dim_name) != rvars.end());
+
+        pair<VarOrRVar, VarOrRVar> split_vars =
+                split_dim(f_handle, vec_var, vec_len, "_vi", "_vo",
+                          estimates, sched);
+
+        f_handle.vectorize(split_vars.first);
+        sched += f_handle.name() + ".vectorize(" +
+                split_vars.first.name() + ");\n";
+
+        if (is_rvar) {
+            rvars.erase(vec_dim_name);
+            rvars.insert(split_vars.first.name());
+            rvars.insert(split_vars.second.name());
+        }
+
+        // TODO: Reorder vector dim to the inner most if it is the inner
+        // most storage dimension of the func
+        if (vec_dim_index > 0) {
+            user_warning << "Outer dim vectorization of var " << vec_dim_name
+                         << " in function " << f_handle.name() << '\n';
+            //Var inner_var(get_base_name(dims[0].var));
+            //f_handle.reorder(split_vars.first, split_vars.second, inner_var);
+        }
+    }
+}
+
+void reorder_dims(Stage f_handle, Definition def,
+                  map<string, int64_t> strides, string &sched) {
+    // TODO: Do not reorder impure reduction dimensions relative to each other
+    vector<Dim> &dims = def.schedule().dims();
+    vector<pair<string, bool>> order;
+
+    for (int d = 0; d < (int)dims.size() - 1; d++) {
+        internal_assert(strides.find(dims[d].var) != strides.end());
+    }
+
+    // Order dims by spatial locality
+    while (strides.size() > 0) {
+        // Find pure dimension wi
```

Comment:
A bit of a grammar nit, but generally, prefer to have comments with complete, grammatically correct sentences, ending in the proper punctuation. (This is not followed universally in Halide but is what we generally strive for.)